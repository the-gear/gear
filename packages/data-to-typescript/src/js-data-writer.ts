import { ReferentialSerializer } from './referential-serializer';
import { isWithKeys, getPropertyPath, getPropertyAccess } from './utils';
// import { WithProperties } from './primitive-serializer';

export type Exports = { [keys: string]: unknown };

class Ref {
  seen = 1;
  exports: string[] = [];
  name: string;

  constructor(name: string) {
    this.name = name;
  }

  export(name: string): this {
    this.exports.push(name);
    return this;
  }

  ref(): this {
    this.seen++;
    return this;
  }
}

class JsSerializer extends ReferentialSerializer {
  constructor(private writer: JsDataWriter) {
    super();
  }

  serializeRecursion(value: unknown, longPath: PropertyKey[], shortPath: PropertyKey[]): string {
    const ref = this.writer.refs.get(value);
    /* istanbul ignore next */
    if (!ref) throw new Error('Unhandled recursion');
    const name = ref.name;
    const long = getPropertyPath(name, ...longPath);
    const short = getPropertyPath(name, ...shortPath);
    this.writer.recursions.push(`${long} = ${short};`);

    return Array.isArray(value) ? `[ /* ${short} */ ]` : `{ /* ${short} */ }`;
  }

  // serializePropertyValue(
  //   value: unknown,
  //   key: PropertyKey,
  //   parent: WithProperties,
  // ): string | undefined {
  //   const result = super.serializePropertyValue(value, key, parent);
  //   const ref = this.writer.refs.get(value);
  //   if (ref) {
  //     this.setReplace(value, getPropertyPath(ref.name, ...this.parentKeys));
  //   }
  //   return result;
  // }
}

/**
 * Serialize data, replace defined values with reference,
 */
export class JsDataWriter {
  serializer = new JsSerializer(this);
  refs = new Map<unknown, Ref>();
  exports = new Map<string, unknown>();
  recursions: string[] = [];

  constructor(public moduleType: 'ESM' | 'CJS' = 'ESM') {}

  addExports(exports: Exports): this {
    for (const [name, value] of Object.entries(exports)) {
      this.add(name, value, true);
    }
    return this;
  }

  addChildren(name: string, parent: unknown) {
    if (isWithKeys(parent)) {
      for (const [key, value] of Object.entries(parent)) {
        if (isWithKeys(value)) {
          this.add(`${name}$${key}`, value, false);
        }
      }
    }
  }

  add(name: string, value: unknown, isExport: boolean = true): Ref {
    if (isExport) {
      if (this.exports.has(name)) {
        const definedExport = this.exports.get(name);
        if (!Object.is(definedExport, value)) {
          throw new TypeError(`Cannot redeclare export ${name}`);
        }
      } else {
        this.exports.set(name, value);
      }
    }

    const ref = this.refs.get(value);
    if (ref) {
      if (isExport) ref.export(name);
      return ref.ref();
    } else {
      const ref = new Ref(name);
      if (isExport) ref.export(name);
      this.refs.set(value, ref);
      this.addChildren(name, value);

      // ensure that any children will be before its parent
      this.refs.delete(value);
      this.refs.set(value, ref);

      return ref;
    }
  }

  private flushRecursion(code: string[]) {
    if (!this.recursions.length) return;
    code.push(this.recursions.join('\n'));
    this.recursions = [];
  }

  protected writeConst(name: string, value: unknown): string {
    return `const ${name} = ${this.serializer.serialize(value)};`;
  }

  protected writeExportConst(name: string, value: unknown): string {
    let exportStr: string;
    switch (this.moduleType) {
      case 'ESM': {
        exportStr = `export const ${name}`;
        break;
      }
      case 'CJS': {
        exportStr = `const ${name} = module.exports${getPropertyAccess(name)}`;
        break;
      }
      default: {
        /* istanbul ignore next */
        throw new TypeError(`Unknown module type: '${this.moduleType}'.`);
      }
    }
    return `${exportStr} = ${this.serializer.serialize(value)};`;
  }

  toString(): string {
    const code = [];
    for (const [value, ref] of this.refs) {
      const exports = ref.exports;
      if (exports.length) {
        const name = exports[0];
        ref.name = name;
        code.push(this.writeExportConst(name, value));
        this.flushRecursion(code);
        this.serializer.setReplace(value, name);
        for (let i = 1; i < exports.length; i++) {
          code.push(this.writeExportConst(exports[i], value));
        }
        code.push();
      } else {
        if (ref.seen > 1) {
          const name = ref.name;
          code.push(this.writeConst(name, value));
          this.serializer.setReplace(value, name);
        }
      }
      this.flushRecursion(code);
    }
    return code.join('\n');
  }
}
