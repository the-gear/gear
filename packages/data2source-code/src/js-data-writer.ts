import { SimpleDataWriter } from './simple-data-writer';
import { getPropertyPath } from './utils';

export class JsDataWriter extends SimpleDataWriter {
  private name: string | null = null;
  private circulars: string[] = [];
  private refs = new Map<object, string>();

  constructor(public valueSubstitutions: Map<unknown, string> = new Map<unknown, string>()) {
    super();
  }

  public static writeDefinition(name: string, value: unknown): string {
    return new this().writeDefinition(name, value).toString();
  }

  public addSubstitution(value: unknown, subst: string): this {
    this.valueSubstitutions.set(value, subst);
    return this;
  }

  public writeDefinition(name: string, value: unknown, binding: string = 'const'): this {
    this.name = name;
    this.ensureNewline();
    this.writeRaw(`${binding} ${name} = `);
    this.visit(value);
    this.writeRaw(`;`);
    this.name = null;

    if (this.circulars.length) {
      this.ensureNewline();
      this.writeRaw(this.circulars.join('\n'));
      this.circulars = [];
    }

    this.valueSubstitutions.set(value, name);
    for (const [val, path] of this.refs) {
      if (!this.valueSubstitutions.has(val)) {
        this.addSubstitution(val, path);
      }
    }

    this.refs = new Map();
    return this;
  }

  visit(value: unknown): void {
    const replace = this.valueSubstitutions.get(value);
    if (replace) {
      this.writeRaw(replace);
    } else {
      super.visit(value);
    }
  }

  protected ['object'](value: object): void {
    /* istanbul ignore next */
    if (!this.name) throw new Error('Name is not defined!');

    if (value === null) return this.visitNull();
    const isArray = Array.isArray(value);
    const writtenPath = this.refs.get(value);
    const path = getPropertyPath(this.name, this.parentKeys);
    if (writtenPath) {
      if (isArray) {
        this.writeRaw(`[/* ref ${writtenPath} */]`);
      } else {
        this.writeRaw(`{/* ref ${writtenPath} */}`);
      }
      this.circulars.push(`${path} = ${writtenPath}; /*ref*/`);
    } else {
      super['object'](value);
      this.refs.set(value, path);
    }
  }

  protected visitRecursiveProperty(
    value: object,
    _key: PropertyKey,
    _parent: object,
    _isArray: boolean,
  ): void {
    /* istanbul ignore next */
    if (!this.name) throw new Error('Name is not defined!');

    const index = this.parents.indexOf(value);
    const shortPath = getPropertyPath(this.name, this.parentKeys.slice(0, index));
    const longPath = getPropertyPath(this.name, this.parentKeys);
    if (Array.isArray(value)) {
      this.writeRaw(`[/* circular ${shortPath} */]`);
    } else {
      this.writeRaw(`{/* circular ${shortPath} */}`);
    }
    this.writeRaw(',');
    this.circulars.push(`${longPath} = ${shortPath}; /*circ*/`);
  }
}
