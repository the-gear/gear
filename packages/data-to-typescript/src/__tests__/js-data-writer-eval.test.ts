import { JsDataWriter, Exports } from '../js-data-writer';
import { runInNewContext } from 'vm';

const str = 'abc';
const obj = { int: 1, str, nullish: null, undef: undefined };
const arr = [obj, obj, 1, str, null, {}, []];
arr.push(arr);

const rec1 = { rec2: {}, rec3: {} };
const rec2 = { rec1 };
rec1.rec2 = rec2;
rec1.rec3 = rec2;

const child1 = { name: 'child1' };
const child2 = { name: 'child2' };
const collection = {
  children: [child1, child2],
};

const moduleHeader = `
const exports = {};
const module = { exports };
`;

const serialize = (exports: Exports) => new JsDataWriter('CJS').addExports(exports).toString();
const evalModule = (code: string) => runInNewContext(moduleHeader + code + '\nmodule.exports');
const serializeThenEval = (exports: Exports) => evalModule(serialize(exports));

describe('JsDataWriter', () => {
  it('can export primitive values', () => {
    expect(serializeThenEval(obj)).toMatchSnapshot();
  });

  it('can export object', () => {
    expect(serializeThenEval({ obj })).toMatchSnapshot();
  });

  it('can export array', () => {
    expect(serializeThenEval({ arr })).toMatchSnapshot();
  });

  it('can export array', () => {
    expect(serializeThenEval({ obj, arr })).toMatchSnapshot();
  });

  it('can export collection and children', () => {
    expect(serializeThenEval({ collection, child1 })).toMatchSnapshot();
  });

  it('can export recursive object', () => {
    expect(serializeThenEval({ rec1, rec2 })).toMatchSnapshot();
  });

  it('can export recursive object', () => {
    expect(serializeThenEval({ rec1, rec2, recAlias1: rec1, recAlias2: rec2 })).toMatchSnapshot();
  });
});
