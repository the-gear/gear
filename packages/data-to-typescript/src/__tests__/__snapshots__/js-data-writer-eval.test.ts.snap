// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsDataWriter can export array 1`] = `
Object {
  "arr": Array [
    Object {
      "int": 1,
      "nullish": null,
      "str": "abc",
      "undef": undefined,
    },
    Object {
      "int": 1,
      "nullish": null,
      "str": "abc",
      "undef": undefined,
    },
    1,
    "abc",
    null,
    Object {},
    Array [],
    Array [
      Object {
        "int": 1,
        "nullish": null,
        "str": "abc",
        "undef": undefined,
      },
      Object {
        "int": 1,
        "nullish": null,
        "str": "abc",
        "undef": undefined,
      },
      1,
      "abc",
      null,
      Object {},
      Array [],
      [Circular],
    ],
  ],
}
`;

exports[`JsDataWriter can export array 2`] = `
Object {
  "arr": Array [
    Object {
      "int": 1,
      "nullish": null,
      "str": "abc",
      "undef": undefined,
    },
    Object {
      "int": 1,
      "nullish": null,
      "str": "abc",
      "undef": undefined,
    },
    1,
    "abc",
    null,
    Object {},
    Array [],
    Array [
      Object {
        "int": 1,
        "nullish": null,
        "str": "abc",
        "undef": undefined,
      },
      Object {
        "int": 1,
        "nullish": null,
        "str": "abc",
        "undef": undefined,
      },
      1,
      "abc",
      null,
      Object {},
      Array [],
      [Circular],
    ],
  ],
  "obj": Object {
    "int": 1,
    "nullish": null,
    "str": "abc",
    "undef": undefined,
  },
}
`;

exports[`JsDataWriter can export collection and children 1`] = `
Object {
  "child1": Object {
    "name": "child1",
  },
  "collection": Object {
    "children": Array [
      Object {
        "name": "child1",
      },
      Object {
        "name": "child2",
      },
    ],
  },
}
`;

exports[`JsDataWriter can export object 1`] = `
Object {
  "obj": Object {
    "int": 1,
    "nullish": null,
    "str": "abc",
    "undef": undefined,
  },
}
`;

exports[`JsDataWriter can export primitive values 1`] = `
Object {
  "int": 1,
  "nullish": null,
  "str": "abc",
  "undef": undefined,
}
`;

exports[`JsDataWriter can export recursive object 1`] = `
Object {
  "rec1": Object {
    "rec2": Object {
      "rec1": Object {
        "rec2": [Circular],
        "rec3": [Circular],
      },
    },
    "rec3": Object {
      "rec1": Object {
        "rec2": [Circular],
        "rec3": [Circular],
      },
    },
  },
  "rec2": Object {
    "rec1": Object {
      "rec2": [Circular],
      "rec3": [Circular],
    },
  },
}
`;

exports[`JsDataWriter can export recursive object 2`] = `
Object {
  "rec1": Object {
    "rec2": Object {
      "rec1": Object {
        "rec2": [Circular],
        "rec3": [Circular],
      },
    },
    "rec3": Object {
      "rec1": Object {
        "rec2": [Circular],
        "rec3": [Circular],
      },
    },
  },
  "rec2": Object {
    "rec1": Object {
      "rec2": [Circular],
      "rec3": [Circular],
    },
  },
  "recAlias1": Object {
    "rec2": Object {
      "rec1": Object {
        "rec2": [Circular],
        "rec3": [Circular],
      },
    },
    "rec3": Object {
      "rec1": Object {
        "rec2": [Circular],
        "rec3": [Circular],
      },
    },
  },
  "recAlias2": Object {
    "rec1": Object {
      "rec2": [Circular],
      "rec3": [Circular],
    },
  },
}
`;
